crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,446 @@
+                    *file = OPENSSL_strndup(p_file, p_line - 1 - p_file);
+                    *line = (int)line_number;
+                    msg = strchr(p_level, ':') + 1;
+                    if (*msg == ' ')
+                        msg++;
+                }
+            }
+        }
+    }
+    return msg;
+}
+
+
+/*
+ * auxiliary function for incrementally reporting texts via the error queue
+ */
+static void put_error(int lib, const char *func, int reason,
+                      const char *file, int line)
+{
+    ERR_new();
+    ERR_set_debug(file, line, func);
+    ERR_set_error(lib, reason, NULL /* no data here, so fmt is NULL */);
+}
+
+#define ERR_print_errors_cb_LIMIT 4096 /* size of char buf[] variable there */
+#define TYPICAL_MAX_OUTPUT_BEFORE_DATA 100
+#define MAX_DATA_LEN (ERR_print_errors_cb_LIMIT-TYPICAL_MAX_OUTPUT_BEFORE_DATA)
+void ossl_cmp_add_error_txt(const char *separator, const char *txt)
+{
+    const char *file = NULL;
+    int line;
+    const char *func = NULL;
+    const char *data = NULL;
+    int flags;
+    unsigned long err = ERR_peek_last_error();
+
+    if (separator == NULL)
+        separator = "";
+    if (err == 0)
+        put_error(ERR_LIB_CMP, NULL, 0, "", 0);
+
+    do {
+        size_t available_len, data_len;
+        const char *curr = txt, *next = txt;
+        char *tmp;
+
+        ERR_peek_last_error_all(&file, &line, &func, &data, &flags);
+        if ((flags & ERR_TXT_STRING) == 0) {
+            data = "";
+            separator = "";
+        }
+        data_len = strlen(data);
+
+        /* workaround for limit of ERR_print_errors_cb() */
+        if (data_len >= MAX_DATA_LEN
+                || strlen(separator) >= (size_t)(MAX_DATA_LEN - data_len))
+            available_len = 0;
+        else
+            available_len = MAX_DATA_LEN - data_len - strlen(separator) - 1;
+        /* MAX_DATA_LEN > available_len >= 0 */
+
+        if (separator[0] == '\0') {
+            const size_t len_next = strlen(next);
+
+            if (len_next <= available_len) {
+                next += len_next;
+                curr = NULL; /* no need to split */
+            }
+            else {
+                next += available_len;
+                curr = next; /* will split at this point */
+            }
+        } else {
+            while (*next != '\0' && (size_t)(next - txt) <= available_len) {
+                curr = next;
+                next = strstr(curr, separator);
+                if (next != NULL)
+                    next += strlen(separator);
+                else
+                    next = curr + strlen(curr);
+            }
+            if ((size_t)(next - txt) <= available_len)
+                curr = NULL; /* the above loop implies *next == '\0' */
+        }
+        if (curr != NULL) {
+            /* split error msg at curr since error data would get too long */
+            if (curr != txt) {
+                tmp = OPENSSL_strndup(txt, curr - txt);
```

Comment:
odd error handling here

Reply:
The whole function is a non-trivial workaround for a pretty ugly hard-coded limitation of `ERR_print_errors_cb()` on the number of chars in an err queue entry.
What this does is to split longer error text into multiple entries.

Reply:
I think when strdup fails, this function should not continue.

Reply:
What in particular do you refer to by your comment; what do you suggest to do?

Reply:
I think you could either not call ERR_add_error_data
if tmp is NULL, or simply return from this function.
Or would you consider calling ERR_add_error_data with NULL
to be the right thing here?

Reply:
Now I got your point that you meant that OPENSSL_strndup may fail.
I've added a simple return since in this case we cannot do much - further calls to ERR_add_error_data etc. would likely fail due to OOM as well.